spellcheck/platform/linux/spellcheck_linux.cpp

```diff
@@ -1,38 +1,117 @@
+	if (tried) return;
+	tried = true;
+	if (!enchant::loader::do_explicit_linking()) return;
+
+	typedef std::unique_ptr<enchant::Dict> DictPtr;
+	brokerHandle = std::make_unique<enchant::Broker>();
+	personal = brokerHandle->request_pwl_dict("personal.dic");
+	validators.push_back(DictPtr(personal));
+
+	std::set<std::string> langs;
+	brokerHandle->list_dicts([](
+			const char *language,
+			const char *provider,
+			const char *description,
+			const char *filename,
+			void *our_payload) {
+		static_cast<decltype(langs)*>(our_payload)->insert(language);
+	}, &langs);
+	for (const auto &language : langs) {
+		validators.push_back(DictPtr(brokerHandle->request_dict(language)));
+	}
 }
 
-void FillSuggestionList(
-	const QString &wrongWord,
-	std::vector<QString> *optionalSuggestions) {
+}  // anonymous namespace
+
+bool Platform::Spellchecker::IsAvailable() {
+	GetValidators();
+	return validators.size() > 1;
 }
 
-void AddWord(const QString &word) {
+void Platform::Spellchecker::KnownLanguages(std::vector<QString> *langCodes) {
+	GetValidators();
+	langCodes->clear();
+	for (const auto &validator : validators) {
+		if (validator.get() == personal) continue;
+		langCodes->push_back(validator->get_lang().c_str());
+	}
 }
 
-void RemoveWord(const QString &word) {
+bool Platform::Spellchecker::CheckSpelling(const QString &wordToCheck) {
+	GetValidators();
+
+	return ranges::any_of(validators, [&](const auto &validator) {
+		return validator->check(wordToCheck.toStdString());
+	}) || validators.empty();
 }
 
-void IgnoreWord(const QString &word) {
+void Platform::Spellchecker::FillSuggestionList(
+		const QString &wrongWord,
+		std::vector<QString> *optionalSuggestions) {
+	GetValidators();
+
+	optionalSuggestions->clear();
+	for (const auto &validator : validators) {
+		auto suggestions = validator->suggest(wrongWord.toStdString());
+		for (const std::string &replacement : suggestions) {
+			if (!replacement.empty()) {
+				optionalSuggestions->push_back(replacement.c_str());
+			}
+			if (optionalSuggestions->size() >= kMaxSuggestions) {
+				break;
+			}
+		}
+		if (!optionalSuggestions->empty()) {
+			break;
+		}
+	}
+}
+
+void Platform::Spellchecker::AddWord(const QString &word) {
+	GetValidators();
+	if (!personal) return;
+	personal->add(word.toStdString());
 }
 
-bool IsWordInDictionary(const QString &wordToCheck) {
-	return false;
+void Platform::Spellchecker::RemoveWord(const QString &word) {
+	GetValidators();
+	if (!personal) return;
+	personal->remove(word.toStdString());
 }
 
-void CheckSpellingText(
-	const QString &text,
-	MisspelledWords *misspelledWordRanges) {
+void Platform::Spellchecker::IgnoreWord(const QString &word) {
+	// what should it do?
 }
 
-} // namespace Platform::Spellchecker
+bool Platform::Spellchecker::IsWordInDictionary(const QString &wordToCheck) {
+	GetValidators();
+
+	return personal && (personal->is_added(wordToCheck.toStdString())
+			|| personal->is_removed(wordToCheck.toStdString()));
```

Comment:
Nitpick: `wordToCheck.toStdString()` should be moved to `const auto` variable to avoid duplication of converting from `QString` to `std::string`.